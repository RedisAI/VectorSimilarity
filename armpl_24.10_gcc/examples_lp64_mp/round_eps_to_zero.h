#ifndef ROUND_EPS_TO_ZERO_H
#define ROUND_EPS_TO_ZERO_H

/*
 * Arm Performance Libraries version 24.10
 * SPDX-FileCopyrightText: Copyright 2015-2024 Arm Limited and/or its affiliates
 */

#include <float.h>
#include <math.h>

// On some architectures we occasionally see tiny differences between an expected
// answer of 0 and the calculated answer, so we choose not to display these
float round_eps_to_zero_f(float x) {
	return fabs(x) > FLT_EPSILON ? x : 0;
}
double round_eps_to_zero_d(double x) {
	return fabs(x) > DBL_EPSILON ? x : 0;
}

#endif
