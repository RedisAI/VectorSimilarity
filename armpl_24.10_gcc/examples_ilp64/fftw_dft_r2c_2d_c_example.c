/*
 * fftw_dft_r2c_2d: FFT of a 2d real sequence
 *
 * Arm Performance Libraries version 24.10
 * SPDX-FileCopyrightText: Copyright 2015-2024 Arm Limited and/or its affiliates
 */

#include <complex.h>
#include <fftw3.h>
#include <math.h>
#include <stdio.h>

#include "round_eps_to_zero.h"

int main(void) {
#define MMAX 20
#define NMAX 20
	double xx[MMAX * NMAX];
	double x[MMAX * NMAX];
	// The output vector is of size m * (n/2+1) as it is Hermitian
	fftw_complex y[MMAX * (NMAX / 2) + 1];

	printf("ARMPL example: FFT of a 2d real sequence using "
	       "fftw_plan_dft_r2c_2d\n");
	printf("-------------------------------------------------------------------"
	       "\n");
	printf("\n");

	/* The 2d sequence of double data */
	int m = 5;
	int n = 2;
	int ix = 0;
	int i, j;
	for (i = 1; i <= m; i++) {
		for (j = 1; j <= n; j++) {
			x[ix] = 1.0 + cos(j + 2 * i);
			// Copy the values into another array to preserve input
			xx[ix] = x[ix];
			ix++;
		}
	}

	// Initialise a plan for a real-to-complex 2d transform from x->y
	fftw_plan f = fftw_plan_dft_r2c_2d(m, n, x, y, FFTW_ESTIMATE);
	// Initialise a plan for a complex-to-real 2d transform from y->x (inverse)
	fftw_plan b = fftw_plan_dft_c2r_2d(m, n, y, x, FFTW_ESTIMATE);

	// Execute the forward plan and then deallocate the plan
	/* NOTE: FFTW does NOT compute a normalised transform -
	 * returned array will contain unscaled values */
	fftw_execute(f);
	fftw_destroy_plan(f);

	printf("Components of discrete Fourier transform:\n");
	printf("\n");
	ix = 0;
	for (i = 0; i < m; i++) {
		for (j = 0; j <= n / 2; j++) {
			// Scale factor of 1/sqrt(m*n) to output normalised data
			double scale = 1.0 / sqrt(m * n);
			double y_real = round_eps_to_zero_d(creal(y[ix]) * scale);
			double y_imag = round_eps_to_zero_d(cimag(y[ix]) * scale);
			printf("%4d,%d   (%7.4f+%7.4fi)\n", i + 1, j + 1, y_real, y_imag);
			ix++;
		}
	}

	// Execute the reverse plan and then deallocate the plan
	/* NOTE: FFTW does NOT compute a normalised transform -
	 * returned array will contain unscaled values */
	fftw_execute(b);
	fftw_destroy_plan(b);

	printf("\n");
	printf("Original sequence as restored by inverse transform:\n");
	printf("\n");
	printf("       Original  Restored\n");
	ix = 0;
	for (i = 0; i < m; i++) {
		for (j = 0; j < n; j++) {
			// Scale factor of 1/m*n to output normalised data
			double scale = 1.0 / (m * n);
			double xx_ix = round_eps_to_zero_d(xx[ix]);
			double x_ix = round_eps_to_zero_d(x[ix] * scale);
			printf("%4d,%d   %7.4f   %7.4f\n", i + 1, j + 1, xx_ix, x_ix);
			ix++;
		}
	}

	return 0;
}
