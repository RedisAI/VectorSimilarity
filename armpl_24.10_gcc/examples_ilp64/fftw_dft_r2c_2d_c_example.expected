ARMPL example: FFT of a 2d real sequence using fftw_plan_dft_r2c_2d
-------------------------------------------------------------------

Components of discrete Fourier transform:

   1,1   ( 2.9430+ 0.0000i)
   1,2   (-0.3241+ 0.0000i)
   2,1   (-0.0232+-0.5574i)
   2,2   (-0.4662+-0.2296i)
   3,1   (-1.1666+ 0.6358i)
   3,2   ( 0.3623+ 0.2619i)
   4,1   (-1.1666+-0.6358i)
   4,2   ( 0.3623+-0.2619i)
   5,1   (-0.0232+ 0.5574i)
   5,2   (-0.4662+ 0.2296i)

Original sequence as restored by inverse transform:

       Original  Restored
   1,1    0.0100    0.0100
   1,2    0.3464    0.3464
   2,1    1.2837    1.2837
   2,2    1.9602    1.9602
   3,1    1.7539    1.7539
   3,2    0.8545    0.8545
   4,1    0.0889    0.0889
   4,2    0.1609    0.1609
   5,1    1.0044    1.0044
   5,2    1.8439    1.8439
